stdlib/asm/mem.masm

```diff
@@ -2,6 +2,118 @@ use.std::crypto::hashes::rpo
 
 # ===== MEMORY FUNCTIONS ==========================================================================
 
+#! Copies `n` felts from `read_ptr` to `write_ptr`.
+#!
+#! If possible, this procedure will copy word-aligned words internally using `memcopy_words` 
+#! procedure, decreasing the total number of cycles required to make a copy.
+#!
+#! It is advised to copy small chunks of memory using `memcopy_felts` procedure instead, since it
+#! won't have a computation logic overhead.
+#!
+#! Inputs:  [n, read_ptr, write_ptr]
+#! Outputs: []
+#!
+#! Where:
+#! - n is the number of felts which should be copied.
+#! - read_ptr is the memory pointer where the felts to copy are stored.
+#! - write_ptr is the memory pointer where the felts will be copied.
+#!
+#! Panics if:
+#! - read_ptr and write_ptr are not a u32 values.
+#!
+#! Total cycles:
+#! - if read and write pointers are mutually unaligned: 23 + 14 * num_felts
+#! - if read and write pointers are mutually aligned, it's equal to the sum of:
+#!   - prologue: 14 cycles
+#!   - prefix handling (worst case): 28 + 14 * num_felts_to_next_wa, where num_felts_to_next_wa is 
+#!     the number of felts required to make the read (and write) pointer word-aligned
+#!   - words handling: 15 + 16 * num_words
+#!   - suffix handling: 9 + 14 * num_remaining_felts
+#!     - in total: 66 + 14 * num_prefix_felts + 16 * num_words + 14 * num_suffix_felts
+export.memcopy
+    # check whether the difference between `read_ptr` and `write_ptr` is 4, otherwise these pointers 
+    # will never be simultaneously word-alined, so we won't be able to use `memcopy_words`
+    movdn.2 u32assert2.err="memory addresses must be valid u32 values" # 2 cycles
+    # => [read_ptr, write_ptr, n]
+
+    # Compute the remainders of dividing read_ptr and write_ptr by 4. Compare them to get the 
+    # "alignment flag".
+    dup.1 u32mod.4 dup.1 u32mod.4 dup.1 eq # 12 cycles
+    # => [alignment_flag, write_ptr%4, read_ptr, write_ptr, n]
+
+    if.true
+        # if read pointer and write pointer are word-aligned between each other, we will be able to
```

Comment:
I think a better terminology here is `mutually word-aligned`, which is defined as `r_ptr - w_ptr = 0 mod 4`